File: apps/macos/SafeWorld/UpdateService.swift
import AppKit
import Foundation
import SafeWorldCore

/// Checks for a newer build, downloads its disk image, and opens it.
///
/// Stops at "the .dmg is mounted and showing", deliberately. Replacing a
/// running app bundle in place is what Sparkle exists to do, and doing it
/// safely means a signed, notarized build so the replacement can be verified —
/// this app is ad-hoc signed (see the README), so a self-replacing updater here
/// would be swapping the binary with nothing checking what it swapped in.
/// Mounting the image and letting the user drag it to Applications keeps
/// Gatekeeper in the loop and is one step for them.
@MainActor
final class UpdateService: ObservableObject {
    enum State: Equatable {
        case idle
        case checking
        /// Checked, and this build is the newest published one.
        case upToDate
        case available(AvailableUpdate)
        /// `fraction` is 0...1, or nil while the size is still unknown.
        case downloading(AvailableUpdate, fraction: Double?)
        /// Downloaded and handed to Finder.
        case downloaded(AvailableUpdate)
        case failed(String)
    }

    @Published private(set) var state: State = .idle

    /// The running build's `CFBundleShortVersionString`, shown in the menu and
    /// compared against the newest release.
    let installedVersion: String = {
        Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0.0.0"
    }()

    private var lastCheckedAt: Date?

    // MARK: Checking

    /// Looks for a newer release. `force` runs even when one ran recently — the
    /// throttle exists so reopening the menu doesn't spend the unauthenticated
    /// rate limit, but a button the user pressed should always do something.
    func check(force: Bool = false) {
        if case .checking = state { return }
        if case .downloading = state { return }
        if !force, let last = lastCheckedAt,
           Date().timeIntervalSince(last) < ReleaseChecker.checkIntervalHours * 3600 {
            return
        }

        state = .checking
        Task {
            do {
                let release = try await fetchLatest()
                lastCheckedAt = Date()
                let update = ReleaseChecker.select(
                    release: release,
                    currentVersion: installedVersion,
                    assetExtension: ".dmg"
                )
                state = update.map(State.available) ?? .upToDate
            } catch {
                // Network, HTTP, and malformed JSON all mean the same thing to
                // the user: we couldn't ask right now.
                state = .failed("Couldn't check for updates.")
            }
        }
    }

    private func fetchLatest() async throws -> GitHubRelease {
        var request = URLRequest(url: ReleaseChecker.latestReleaseURL)
        request.cachePolicy = .reloadIgnoringLocalCacheData
        request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        // GitHub rejects API requests that don't identify themselves.
        request.setValue("SafeWorld-macOS", forHTTPHeaderField: "User-Agent")

        let (data, response) = try await URLSession.shared.data(for: request)
        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            throw URLError(.badServerResponse)
        }
        return try ReleaseChecker.parse(data)
    }

    // MARK: Downloading

    /// Downloads the disk image to ~/Downloads and opens it. A release with no
    /// .dmg attached (the .apk and .exe live in the same release) still has a
    /// web page worth going to, so that case opens the page instead of failing.
    func download(_ update: AvailableUpdate) {
        guard let assetURL = update.assetURL else {
            NSWorkspace.shared.open(update.pageURL)
            return
        }
        if case .downloading = state { return }

        state = .downloading(update, fraction: nil)
        Task {
            do {
                let file = try await downloadToDownloadsFolder(assetURL, update: update)
                state = .downloaded(update)
                // Mounts the image and brings its window up, which is where the
                // drag-to-Applications step happens.
                NSWorkspace.shared.open(file)
            } catch {
                state = .failed("The download didn't finish.")
            }
        }
    }

    func openReleasePage(_ update: AvailableUpdate) {
        NSWorkspace.shared.open(update.pageURL)
    }

    private func downloadToDownloadsFolder(
        _ url: URL,
        update: AvailableUpdate
    ) async throws -> URL {
        let (bytes, response) = try await URLSession.shared.bytes(from: url)
        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            throw URLError(.badServerResponse)
        }
        // The release metadata's size is the more reliable of the two: the asset
        // download redirects, and the final response doesn't always carry a length.
        let total = update.sizeBytes > 0 ? update.sizeBytes : response.expectedContentLength

        var data = Data()
        if total > 0 { data.reserveCapacity(Int(total)) }
        var lastReported = 0.0

        for try await byte in bytes {
            data.append(byte)
            guard total > 0 else { continue }
            let fraction = Double(data.count) / Double(total)
            // Republishing per byte would repaint the menu tens of thousands of
            // times; 1% steps are past what the eye resolves on a short bar.
            if fraction - lastReported >= 0.01 {
                lastReported = fraction
                state = .downloading(update, fraction: min(fraction, 1))
            }
        }

        let downloads = try FileManager.default.url(
            for: .downloadsDirectory, in: .userDomainMask, appropriateFor: nil, create: true
        )
        let target = downloads.appendingPathComponent(url.lastPathComponent)
        // Replace a previous attempt rather than piling up "file-1.dmg".
        try? FileManager.default.removeItem(at: target)
        try data.write(to: target)
        return target
    }
}

File: apps/windows/SafeWorld.App/UpdateService.cs
using System.Diagnostics;
using System.Reflection;
using SafeWorld.Core;

namespace SafeWorld.App;

/// <summary>
/// Checks for a newer build, downloads its .exe, and hands it to the user — the Windows half of
/// macOS's <c>UpdateService.swift</c> and Android's <c>UpdateManager.kt</c>.
/// <para>
/// <b>Windows was the one platform with no updater at all.</b> Android, iOS, and macOS have had
/// one since 0.2.0; this app has been shipping as a downloaded .exe with no way of knowing it was
/// six releases stale, which is the worst case for a blocklist product — the app keeps working, so
/// nothing ever prompts the user to notice.
/// </para>
/// <para>
/// <b>It stops at "the installer is downloaded and revealed", deliberately.</b> The release is a
/// self-contained single-file .exe, so updating means replacing the very file this process is
/// running from — Windows holds an exclusive lock on it, so the swap has to be done by something
/// that outlives us: a detached script that waits for the process to exit, moves the new file over
/// the old, and relaunches. That is a real technique, and it is also the technique that leaves a
/// user with no app at all if it dies between the delete and the move. This app is unsigned (see
/// the README), so there is nothing to verify the replacement against either. macOS made the same
/// call for the same reason and stops at mounting the .dmg.
/// </para>
/// </summary>
internal sealed class UpdateService
{
    internal enum Status { Idle, Checking, UpToDate, Available, Downloading, Downloaded, Failed }

    /// <summary>Raised whenever <see cref="State"/> changes, so the tray menu can redraw.</summary>
    public event Action? Changed;

    public Status State { get; private set; } = Status.Idle;
    public AvailableUpdate? Update { get; private set; }
    /// <summary>0..100 while downloading, or -1 while the size is still unknown.</summary>
    public int Percent { get; private set; } = -1;
    public string? Message { get; private set; }

    /// <summary>
    /// The running build's assembly version, shown in the menu and compared against the newest
    /// release. Trimmed to three components: the assembly version is always stamped as four
    /// ("0.5.0.0") and comparing that against the tag "0.5.0" is only equal by luck of the
    /// missing-components-are-zero rule — better to compare what we mean.
    /// </summary>
    public static string InstalledVersion { get; } = ReadInstalledVersion();

    private static string ReadInstalledVersion()
    {
        var raw = Assembly.GetExecutingAssembly().GetName().Version;
        return raw is null ? "0.0.0" : $"{raw.Major}.{raw.Minor}.{raw.Build}";
    }

    private static readonly HttpClient Http = CreateClient();

    private static HttpClient CreateClient()
    {
        var client = new HttpClient();
        client.DefaultRequestHeaders.Add("Accept", "application/vnd.github+json");
        // GitHub rejects API requests that don't identify themselves.
        client.DefaultRequestHeaders.Add("User-Agent", "SafeWorld-Windows");
        return client;
    }

    private DateTime? _lastCheckedAt;

    private void Set(Status state, string? message = null)
    {
        State = state;
        Message = message;
        Changed?.Invoke();
    }

    // MARK: Checking

    /// <summary>
    /// Looks for a newer release. <paramref name="force"/> runs even when one ran recently — the
    /// throttle exists so reopening the tray menu doesn't spend the unauthenticated rate limit,
    /// but a button the user pressed should always do something.
    /// </summary>
    public async Task CheckAsync(bool force = false)
    {
        if (State is Status.Checking or Status.Downloading) return;
        if (!force && _lastCheckedAt is { } last
            && (DateTime.UtcNow - last).TotalHours < ReleaseChecker.CheckIntervalHours)
        {
            return;
        }

        Set(Status.Checking);
        try
        {
            var json = await Http.GetStringAsync(ReleaseChecker.LatestReleaseUrl);
            var release = ReleaseChecker.Parse(json);
            _lastCheckedAt = DateTime.UtcNow;

            Update = release is null
                ? null
                : ReleaseChecker.Select(release, InstalledVersion, ReleaseChecker.AssetExtension);
            Set(Update is null ? Status.UpToDate : Status.Available);
        }
        catch (Exception)
        {
            // Network, HTTP, and malformed JSON all mean the same thing to the user: we couldn't
            // ask right now. Never surfaced as "no update", which would be a wrong answer rather
            // than an absent one.
            Set(Status.Failed, "Couldn't check for updates.");
        }
    }

    // MARK: Downloading

    /// <summary>
    /// Downloads the installer to the Downloads folder and reveals it in Explorer. A release with
    /// no .exe attached (the .apk and .dmg live in the same release) still has a web page worth
    /// going to, so that case opens the page instead of failing.
    /// </summary>
    public async Task DownloadAsync(AvailableUpdate update)
    {
        if (update.AssetUrl is null)
        {
            OpenInBrowser(update.PageUrl);
            return;
        }
        if (State == Status.Downloading) return;

        Update = update;
        Percent = -1;
        Set(Status.Downloading);

        try
        {
            var path = await DownloadToDownloadsAsync(update.AssetUrl, update.SizeBytes);
            Set(Status.Downloaded);
            RevealInExplorer(path);
        }
        catch (Exception)
        {
            Set(Status.Failed, "The download didn't finish.");
        }
    }

    private async Task<string> DownloadToDownloadsAsync(string url, long knownSize)
    {
        using var response = await Http.GetAsync(url, HttpCompletionOption.ResponseHeadersRead);
        response.EnsureSuccessStatusCode();

        // The release metadata's size is the more reliable of the two: the asset download
        // redirects, and the final response doesn't always carry a length.
        var total = knownSize > 0 ? knownSize : response.Content.Headers.ContentLength ?? 0;

        var target = Path.Combine(DownloadsFolder(), Path.GetFileName(new Uri(url).LocalPath));
        // Written to a temporary name and moved into place, so an interrupted download cannot
        // leave something that looks like a complete installer sitting in Downloads.
        var partial = target + ".part";

        await using (var source = await response.Content.ReadAsStreamAsync())
        await using (var file = File.Create(partial))
        {
            var buffer = new byte[81920];
            long read = 0;
            var lastReported = -1;
            int n;
            while ((n = await source.ReadAsync(buffer)) > 0)
            {
                await file.WriteAsync(buffer.AsMemory(0, n));
                read += n;
                if (total <= 0) continue;

                var percent = (int)(read * 100 / total);
                // Rebuilding the menu per chunk would repaint it thousands of times for a 140 MB
                // download; whole percent steps are past what the eye resolves anyway.
                if (percent == lastReported) continue;
                lastReported = percent;
                Percent = Math.Min(percent, 100);
                Changed?.Invoke();
            }
        }

        // Replace a previous attempt rather than piling up "file (1).exe".
        if (File.Exists(target)) File.Delete(target);
        File.Move(partial, target);
        return target;
    }

    private static string DownloadsFolder()
    {
        // There is no Environment.SpecialFolder for Downloads. The profile-relative path is what
        // the shell uses by default, and falling back to the profile root is better than throwing
        // for someone who has relocated it.
        var profile = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        var downloads = Path.Combine(profile, "Downloads");
        return Directory.Exists(downloads) ? downloads : profile;
    }

    private static void RevealInExplorer(string path)
    {
        // Selects the file rather than just opening the folder, so it is obvious which of the
        // user's downloads is the one that was just fetched.
        Process.Start(new ProcessStartInfo("explorer.exe", $"/select,\"{path}\"") { UseShellExecute = true });
    }

    public static void OpenInBrowser(string url)
    {
        Process.Start(new ProcessStartInfo(url) { UseShellExecute = true });
    }
}
